### PR TITLE
implement `call dolt_conflicts_resolve()`

### PR DESCRIPTION
**Pull request by JCOR11599**
Closes https://github.com/dolthub/dolt/issues/2475

### CODE REVIEW

File: go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+	}
+
+	// swap the maps when resolving with theirs
+	if !ours {
+		ourMap, theirMap = theirMap, ourMap
+	}
+
+	// resolve conflicts with left
+	var indexEdits []tree.Diff
+	merged, _, err := prolly.MergeMaps(ctx, ourMap, theirMap, baseMap, func(left, right tree.Diff) (tree.Diff, bool) {
```
Comment (druvv) [inline go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go:139]: Redoing the merge to resolve conflicts is going to be time consuming... we should iter the conflicts and take either the ours or theirs values.

File: go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+	baseMap := durable.ProllyMapFromIndex(baseIdx)
+
+	theirMap, err := getProllyRowMaps(ctx, tbl.ValueReadWriter(), tbl.NodeStore(), cnfArt.TheirRootIsh, tblName)
+	if err != nil {
+		return nil, err
+	}
+
+	// swap the maps when resolving with theirs
+	if !ours {
+		ourMap, theirMap = theirMap, ourMap
```
Comment (druvv) [inline go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go:134]: dolt conflicts resolve --ours on the cli just deletes all the conflicts. When a conflict occurs, the ours value is retained in the working set so no work needs to be done. 

I would check if `ours` is true earlier in this function our prior to it and just delete the conflicts to make that behavior explicit.

Comment (druvv) [reply go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go:134]: See here:
https://github.com/dolthub/dolt/blob/7ba66fdc762d0274e4d74bb02dd8d6e7148df247/go/cmd/dolt/commands/cnfcmds/auto_resolve.go#L192-L192

File: go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+		cnfArt, err := iter.Next(ctx)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		if once {
+			theirMap, err = getProllyRowMaps(ctx, tbl.ValueReadWriter(), tbl.NodeStore(), cnfArt.TheirRootIsh, tblName)
```
Comment (druvv) [inline go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go:119]: |TheirRootIsh| may not be the same for each conflict. You may have to fetch the table more than once. I would store the table outside the loop along with the rootish, if the rootish is different than the one we loaded the table from, reload the table.

File: go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+		return nil, err
+	}
+	mutIdxs, err := merge.GetMutableSecondaryIdxs(ctx, sch, idxSet)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: could do this inside the loop while resolving row data
+	// Update secondary indexes
+	for _, mutIdx := range mutIdxs {
```
Comment (druvv) [inline go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go:182]: yea do this inside the conflict iteration loop, not good to hold the index edits in memory. Also, I wouldn't write code to load the tuples into a tree.Diff only to check if `From` or `To` is empty.

**Review by druvv — COMMENTED** [review]
What about keyless tables in the new format?

**Comment by druvv** [comment]
> What about keyless tables in the new format?

Looks like it works! Nevermind.

**Review by druvv — APPROVED** [review]
LGTM. I think the team prefers engine tests over bats tests especially when writing sql functionality.

Last time, I wrote most of the tests in engine tests and then added a smoke test in bats.

File: go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+	"github.com/dolthub/dolt/go/store/prolly/tree"
+	"github.com/dolthub/dolt/go/store/types"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+var ErrConfSchIncompatible = errors.New("the conflict schema's columns are not equal to the current schema's columns, please resolve manually")
+
+// DoltConflictsCatFunc runs a `dolt commit` in the SQL context, committing staged changes to head.
+// Deprecated: please use the version in the dprocedures package
+type DoltConflictsCatFunc struct {
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go:49]: we're not exposing a sql function, only a sql procedure, so just move this stuff into `dprocedures`

File: go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+	tbls := apr.Args
+	if len(tbls) == 1 && tbls[0] == "." {
+		if allTables, err := root.TablesInConflict(ctx); err != nil {
+			return 1, err
+		} else {
+			tbls = allTables
+		}
+	}
+
+	dEnv := env.Load(ctx, env.GetCurrentUserHomeDir, filesys.LocalFS, doltdb.LocalDirDoltDB, "")
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go:450]: the dolt environment should only be created at startup. get the editor options from the session:

```diff
diff --git a/go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go b/go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go
index 0dea4b326..743c26816 100644
--- a/go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go
+++ b/go/libraries/doltcore/sqle/dfunctions/dolt_conflicts_resolve.go
@@ -26,14 +26,12 @@ import (
 	"github.com/dolthub/dolt/go/libraries/doltcore/conflict"
 	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
 	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb/durable"
-	"github.com/dolthub/dolt/go/libraries/doltcore/env"
 	"github.com/dolthub/dolt/go/libraries/doltcore/merge"
 	"github.com/dolthub/dolt/go/libraries/doltcore/row"
 	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
 	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
 	"github.com/dolthub/dolt/go/libraries/doltcore/table"
 	"github.com/dolthub/dolt/go/libraries/doltcore/table/editor"
-	"github.com/dolthub/dolt/go/libraries/utils/filesys"
 	"github.com/dolthub/dolt/go/libraries/utils/set"
 	"github.com/dolthub/dolt/go/store/hash"
 	"github.com/dolthub/dolt/go/store/prolly"
@@ -205,13 +203,8 @@ func resolveProllyConflicts(ctx *sql.Context, tbl *doltdb.Table, tblName string,
 	return newTbl, nil
 }
 
-func resolvePkConflicts(ctx *sql.Context, dEnv *env.DoltEnv, tbl *doltdb.Table, tblName string, sch schema.Schema, conflicts types.Map) (*doltdb.Table, error) {
+func resolvePkConflicts(ctx *sql.Context, opts editor.Options, tbl *doltdb.Table, tblName string, sch schema.Schema, conflicts types.Map) (*doltdb.Table, error) {
 	// Create table editor
-	tmpDir, err := dEnv.TempTableFilesDir()
-	if err != nil {
-		return nil, err
-	}
-	opts := editor.Options{Deaf: dEnv.DbEaFactory(), Tempdir: tmpDir}
 	tblEditor, err := editor.NewTableEditor(ctx, tbl, sch, tblName, opts)
 	if err != nil {
 		return nil, err
@@ -307,7 +300,7 @@ func resolveKeylessConflicts(ctx *sql.Context, tbl *doltdb.Table, conflicts type
 	return tbl.UpdateNomsRows(ctx, rowData)
 }
 
-func resolveNomsConflicts(ctx *sql.Context, dEnv *env.DoltEnv, tbl *doltdb.Table, tblName string, sch schema.Schema) (*doltdb.Table, error) {
+func resolveNomsConflicts(ctx *sql.Context, opts editor.Options, tbl *doltdb.Table, tblName string, sch schema.Schema) (*doltdb.Table, error) {
 	// Get conflicts
 	_, confIdx, err := tbl.GetConflicts(ctx)
 	if err != nil {
@@ -319,7 +312,7 @@ func resolveNomsConflicts(ctx *sql.Context, dEnv *env.DoltEnv, tbl *doltdb.Table
 		return resolveKeylessConflicts(ctx, tbl, conflicts)
 	}
 
-	return resolvePkConflicts(ctx, dEnv, tbl, tblName, sch, conflicts)
+	return resolvePkConflicts(ctx, opts, tbl, tblName, sch, conflicts)
 }
 
 func validateConstraintViolations(ctx *sql.Context, before, after *doltdb.RootValue, table string) error {
@@ -352,7 +345,7 @@ func clearTableAndUpdateRoot(ctx *sql.Context, root *doltdb.RootValue, tbl *dolt
 	return newRoot, nil
 }
 
-func ResolveConflicts(ctx *sql.Context, dEnv *env.DoltEnv, dSess *dsess.DoltSession, root *doltdb.RootValue, dbName string, ours bool, tblNames []string) error {
+func ResolveConflicts(ctx *sql.Context, dSess *dsess.DoltSession, root *doltdb.RootValue, dbName string, ours bool, tblNames []string) error {
 	for _, tblName := range tblNames {
 		tbl, ok, err := root.GetTable(ctx, tblName)
 		if err != nil {
@@ -383,11 +376,17 @@ func ResolveConflicts(ctx *sql.Context, dEnv *env.DoltEnv, dSess *dsess.DoltSess
 			return ErrConfSchIncompatible
 		}
 
+		state, _, err := dSess.LookupDbState(ctx, dbName)
+		if err != nil {
+			return err
+		}
+		opts := state.WriteSession.GetOptions()
+
 		if !ours {
 			if tbl.Format() == types.Format_DOLT {
 				tbl, err = resolveProllyConflicts(ctx, tbl, tblName, sch)
 			} else {
-				tbl, err = resolveNomsConflicts(ctx, dEnv, tbl, tblName, sch)
+				tbl, err = resolveNomsConflicts(ctx, opts, tbl, tblName, sch)
 			}
 			if err != nil {
 				return err
@@ -447,8 +446,7 @@ func DoDoltConflictsResolve(ctx *sql.Context, args []string) (int, error) {
 		}
 	}
 
-	dEnv := env.Load(ctx, env.GetCurrentUserHomeDir, filesys.LocalFS, doltdb.LocalDirDoltDB, "")
-	err = ResolveConflicts(ctx, dEnv, dSess, root, dbName, ours, tbls)
+	err = ResolveConflicts(ctx, dSess, root, dbName, ours, tbls)
 	if err != nil {
 		return 1, err
 	}
```

**Review by andrew-wm-arthur — APPROVED** [review]
Looks good, couple comments

**Pull request merged**